File: src/codeforces/set0/set0/set060/set060/d/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	fmt.Println(drive(reader))
}

// maxA: problem bound on a[i]. maxHyp: hypotenuse can exceed maxA when both legs are ≤ maxA.
const maxA = 10_000_000
const maxHyp = 15_000_000

// posScratch[value] = lawn index + 1, or 0 if absent. One buffer avoids reallocating ~40 MiB per run/test.
var posScratch []int32

func ensurePosScratch() {
	if posScratch == nil {
		posScratch = make([]int32, maxA+1)
	}
}

func drive(reader *bufio.Reader) int {
	var n int
	fmt.Fscan(reader, &n)
	a := make([]int, n)
	for i := range n {
		fmt.Fscan(reader, &a[i])
	}
	return solve(a)
}

func gcd(a, b int) int {
	for b > 0 {
		a, b = b, a%b
	}
	return a
}

func solve(a []int) int {
	n := len(a)
	ensurePosScratch()
	pos := posScratch
	for i := range a {
		pos[a[i]] = int32(i + 1)
	}
	defer func() {
		for i := range a {
			pos[a[i]] = 0
		}
	}()

	set := NewDSU(n)
	enumeratePrimitiveTriples(pos, set)

	res := 0
	for i := range n {
		if set.Find(i) == i {
			res++
		}
	}
	return res
}

func enumeratePrimitiveTriples(pos []int32, set *DSU) {
	uniteVals := func(u, v int) {
		pu := pos[u]
		pv := pos[v]
		if pu != 0 && pv != 0 {
			set.Union(int(pu-1), int(pv-1))
		}
	}

	for m := 2; m*m+1 <= maxHyp; m++ {
		for k := 1; k < m && m*m+k*k <= maxHyp; k++ {
			if gcd(m, k) != 1 || (m^k)&1 == 0 {
				continue
			}
			legA := m*m - k*k
			legB := 2 * m * k
			hyp := m*m + k*k
			l1, l2 := legA, legB
			if l1 > l2 {
				l1, l2 = l2, l1
			}
			if l1 <= 0 || l1 > maxA || l2 > maxA {
				continue
			}
			if gcd(l1, l2) != 1 || gcd(l1, hyp) != 1 || gcd(l2, hyp) != 1 {
				continue
			}
			uniteVals(l1, l2)
			if hyp <= maxA {
				uniteVals(l1, hyp)
				uniteVals(l2, hyp)
			}
		}
	}
}

type DSU struct {
	arr []int
	cnt []int
}

func NewDSU(n int) *DSU {
	arr := make([]int, n)
	cnt := make([]int, n)
	for i := range n {
		arr[i] = i
		cnt[i] = 1
	}
	return &DSU{arr, cnt}
}

func (d *DSU) Find(x int) int {
	if d.arr[x] != x {
		d.arr[x] = d.Find(d.arr[x])
	}
	return d.arr[x]
}

func (d *DSU) Union(x, y int) bool {
	px := d.Find(x)
	py := d.Find(y)
	if px == py {
		return false
	}
	if d.cnt[px] < d.cnt[py] {
		px, py = py, px
	}
	d.cnt[px] += d.cnt[py]
	d.arr[py] = px
	return true
}

File: src/codeforces/set0/set0/set060/set060/d/cpp/solution.cpp
// Codeforces 60D — Mushrooms (memory-friendly).
// No per-value adjacency vectors: O(maxA) for pos[] + O(n) DSU.
// Build: g++ -std=c++17 -O2 -pipe -o solution solution.cpp
// (On Codeforces Linux you may add -static -s if you want a static stripped binary.)

#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>

using namespace std;

static constexpr int kMaxA = 10'000'000;
// Hypotenuse when both legs ≤ kMaxA is ≤ ⌈√2·kMaxA⌉; keep a small margin.
static constexpr int kMaxHyp = 15'000'000;

static int gcd(int a, int b) {
    while (b != 0) {
        int t = a % b;
        a = b;
        b = t;
    }
    return a;
}

struct DSU {
    vector<int> parent;
    vector<unsigned char> rankv;

    explicit DSU(int n) : parent(n), rankv(n, 0) {
        iota(parent.begin(), parent.end(), 0);
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    void unite(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) {
            return;
        }
        if (rankv[a] < rankv[b]) {
            swap(a, b);
        }
        parent[b] = a;
        if (rankv[a] == rankv[b]) {
            ++rankv[a];
        }
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) {
        return 0;
    }

    vector<int> pos(static_cast<size_t>(kMaxA) + 1, 0);
    for (int i = 0; i < n; ++i) {
        int x;
        cin >> x;
        pos[static_cast<size_t>(x)] = i + 1; // 0 = absent
    }

    DSU dsu(n);

    auto unite_vals = [&](int u, int v) {
        int pu = pos[static_cast<size_t>(u)];
        int pv = pos[static_cast<size_t>(v)];
        if (pu != 0 && pv != 0) {
            dsu.unite(pu - 1, pv - 1);
        }
    };

    for (int m = 2; m * m + 1 <= kMaxHyp; ++m) {
        for (int k = 1; k < m && m * m + k * k <= kMaxHyp; ++k) {
            if (gcd(m, k) != 1 || ((m ^ k) & 1) == 0) {
                continue;
            }
            int leg_a = m * m - k * k;
            int leg_b = 2 * m * k;
            int hyp = m * m + k * k;
            int l1 = min(leg_a, leg_b);
            int l2 = max(leg_a, leg_b);
            if (l1 <= 0 || l1 > kMaxA || l2 > kMaxA) {
                continue;
            }
            if (gcd(l1, l2) != 1 || gcd(l1, hyp) != 1 || gcd(l2, hyp) != 1) {
                continue;
            }
            unite_vals(l1, l2);
            if (hyp <= kMaxA) {
                unite_vals(l1, hyp);
                unite_vals(l2, hyp);
            }
        }
    }

    int ans = 0;
    for (int i = 0; i < n; ++i) {
        if (dsu.find(i) == i) {
            ++ans;
        }
    }
    cout << ans << '\n';
    return 0;
}
